cgen error when returning an Option/Result from an `or {}` block

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
V full version: V 0.4.4 5306469.7e5ffb7
OS: windows, Microsoft Windows 10 Pro v19045 64-bit
Processor: 20 cpus, 64bit, little endian, 

getwd: C:\Users\user\Desktop\Code\leaflang
vexe: C:\Users\user\Desktop\Code\v\v.exe
vexe mtime: 2024-01-28 08:36:51

vroot: OK, value: C:\Users\user\Desktop\Code\v
VMODULES: OK, value: C:\Users\user\.vmodules
VTMP: OK, value: C:\Users\user\AppData\Local\Temp\v_0

Git version: git version 2.42.0.windows.1
Git vroot status: weekly.2024.03-91-g7e5ffb71
.git/config present: true

CC version: Error: 'cc' is not recognized as an internal or external command,

operable program or batch file.


thirdparty/tcc status: thirdparty-windows-amd64 a39eb79b

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg test.v`
```v
fn find_startswith_string(a []string, search string) ?string {
	for s in a {
		if s.starts_with(search) {
			return s
		}
	}
	return none
}

fn find_any_startswith_string(a []string, b []string, search string) ?string {
	// cannot convert 'struct _option_string' to 'struct string'
	// V wants the or {} block to return a string, but find_startswith_string returns ?string
 	return find_startswith_string(a, search) or { find_startswith_string(b, search) }
}

fn main() {
	deadbeef := find_any_startswith_string(['foobar', 'barfoo'], ['deadbeef', 'beefdead'],
		'dead')
	assert deadbeef or { panic('unreachable') } == 'deadbeef'
}

```

**What did you expect to see?**

Successful compilation, assert succeeds


**What did you see instead?**
```
==================
C:/Users/user/AppData/Local/Temp/v_0/test.01HN7KE7PZSE8B70ZE078KJQ15.tmp.c:6861: warning: implicit declaration of function 'tcc_backtrace'
C:/Users/user/AppData/Local/Temp/v_0/test.01HN7KE7PZSE8B70ZE078KJQ15.tmp.c:13018: error: cannot convert 'struct _option_string' to 'struct string'
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```

Relevant generated C code:

```c
VV_LOCAL_SYMBOL _option_string main__find_any_startswith_string(Array_string a, Array_string b, string search) {
	_option_string _t2 = main__find_startswith_string(a, search);
	if (_t2.state != 0) {
		IError err = _t2.err;
		*(string*) _t2.data = main__find_startswith_string(b, search);
	}
	
 	_option_string _t1;
	_option_ok(&(string[]) {  (*(string*)_t2.data) }, (_option*)(&_t1), sizeof(string));
	return _t1;
}
```

![image](https://github.com/vlang/v/assets/108888572/41daeabb-b969-4b0c-99bd-f89a9e60d9c5)

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -6622,19 +6622,29 @@ fn (mut g Gen) gen_or_block_stmts(cvar_name string, cast_typ string, stmts []ast
 						is_array_fixed = expr_stmt.expr is ast.ArrayInit
 							&& g.table.final_sym(return_type).kind == .array_fixed
 						if !is_array_fixed {
-							g.write('*(${cast_typ}*) ${cvar_name}.data = ')
+							if g.inside_return && expr_stmt.expr is ast.CallExpr
+								&& return_type.has_option_or_result() {
+								g.write('${cvar_name} = ')
+							} else {
+								g.write('*(${cast_typ}*) ${cvar_name}.data = ')
+							}
 						}
 					} else {
 						g.write('${cvar_name} = ')
 					}
-
 					if is_array_fixed {
 						g.write('memcpy(${cvar_name}.data, (${cast_typ})')
 					}
-					old_inside_opt_data := g.inside_opt_data
-					g.inside_opt_data = true
-					g.expr_with_cast(expr_stmt.expr, expr_stmt.typ, return_type.clear_option_and_result())
-					g.inside_opt_data = old_inside_opt_data
+					// return expr or { fn_returns_option() }
+					if is_option && g.inside_return && expr_stmt.expr is ast.CallExpr
+						&& return_type.has_option_or_result() {
+						g.expr_with_cast(expr_stmt.expr, expr_stmt.typ, return_type)
+					} else {
+						old_inside_opt_data := g.inside_opt_data
+						g.inside_opt_data = true
+						g.expr_with_cast(expr_stmt.expr, expr_stmt.typ, return_type.clear_option_and_result())
+						g.inside_opt_data = old_inside_opt_data
+					}
 					if is_array_fixed {
 						g.write(', sizeof(${cast_typ}))')
 					}

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -6622,7 +6622,8 @@ fn (mut g Gen) gen_or_block_stmts(cvar_name string, cast_typ string, stmts []ast
 						is_array_fixed = expr_stmt.expr is ast.ArrayInit
 							&& g.table.final_sym(return_type).kind == .array_fixed
 						if !is_array_fixed {
-							if g.inside_return && expr_stmt.expr is ast.CallExpr
+							if g.inside_return && !g.inside_struct_init
+								&& expr_stmt.expr is ast.CallExpr
 								&& return_type.has_option_or_result() {
 								g.write('${cvar_name} = ')
 							} else {

--- a/vlib/v/tests/option_call_on_orexpr_test.v
+++ b/vlib/v/tests/option_call_on_orexpr_test.v
@@ -1,0 +1,20 @@
+fn find_startswith_string(a []string, search string) ?string {
+	for s in a {
+		if s.starts_with(search) {
+			return s
+		}
+	}
+	return none
+}
+
+fn find_any_startswith_string(a []string, b []string, search string) ?string {
+	// cannot convert 'struct _option_string' to 'struct string'
+	// V wants the or {} block to return a string, but find_startswith_string returns ?string
+	return find_startswith_string(a, search) or { find_startswith_string(b, search) }
+}
+
+fn test_main() {
+	deadbeef := find_any_startswith_string(['foobar', 'barfoo'], ['deadbeef', 'beefdead'],
+		'dead')
+	assert deadbeef or { panic('unreachable') } == 'deadbeef'
+}

